src/Makefile

```diff
@@ -46,10 +47,6 @@ LUAJIT_LDLIBS := $(shell pkg-config luajit --libs)
 CFLAGS  += $(LUAJIT_CFLAGS)
 LDFLAGS += $(LUAJIT_LDLIBS)
 
-ifeq ($(strip $(LUAJIT)),)
```

Comment:
I don't see how this change prevents anything. Both the existing method and the new one allow overriding `$(LUAJIT)`, and in a cross-compilation scenario it had to be defined before, and it must be defined with the new code as well.

I don't see how relying to a `$PATH` search for `luajit` is any better. Like I said, the build recipes that take care of cross-compilation **must** explicitly override `$(LUAJIT)`. And for that this patch is not necessary.

Reply:
The issue is that no build system (at least no Yocto, neither buildroot) will set the a LUAJIT environment variable. That's not something standard. We have kind of invented such environment variable here.

What Yocto does is to modify the PATH and puts firsts the path where it has build the toolchain (Yocto has this concept of native tools, which are tools that run on the build host like: luajit, qemu, pkg-config, etc). You shouldn't need to rely on your distro packages for that, Yocto will build and use its own tools.
So on a [Yocto recipe for dyz](https://github.com/Igalia/meta-webkit/blob/master/recipes-browser/dyz/dyz_0.1.bb) you simply declare a dependency on luajit-native (because you need the luajit program that runs on the build host) and luajit (because you need to link with the cross-built luajit library for the target system) and you are fine. You don't need to install your distro's luajit package at all, as Yocto will build a luajit binary that runs on your host.

Example:
```
$  bitbake -e dyz|grep "export PATH="
export PATH="/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots-uninative/x86_64-linux/usr/bin:/home/igalia/clopez/yocto-test/build_wpe/sources/poky/scripts:/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots/x86_64-linux/usr/bin/arm-poky-linux-gnueabi:/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots/imx6dl-atlash/usr/bin/crossscripts:/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots/x86_64-linux/usr/sbin:/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots/x86_64-linux/usr/bin:/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots/x86_64-linux/sbin:/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots/x86_64-linux/bin:/home/igalia/clopez/yocto-test/build_wpe/sources/poky/scripts:/home/igalia/clopez/yocto-test/build_wpe/sources/poky/bitbake/bin:/usr/lib/ccache:/sbin:/bin:/usr/sbin:/usr/bin:/usr/games:/usr/local/sbin:/usr/local/bin:/home/igalia/clopez/depot_tools"

$ export PATH="/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots-uninative/x86_64-linux/usr/bin:/home/igalia/clopez/yocto-test/build_wpe/sources/poky/scripts:/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots/x86_64-linux/usr/bin/arm-poky-linux-gnueabi:/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots/imx6dl-atlash/usr/bin/crossscripts:/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots/x86_64-linux/usr/sbin:/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots/x86_64-linux/usr/bin:/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots/x86_64-linux/sbin:/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots/x86_64-linux/bin:/home/igalia/clopez/yocto-test/build_wpe/sources/poky/scripts:/home/igalia/clopez/yocto-test/build_wpe/sources/poky/bitbake/bin:/usr/lib/ccache:/sbin:/bin:/usr/sbin:/usr/bin:/usr/games:/usr/local/sbin:/usr/local/bin:/home/igalia/clopez/depot_tools"
$ which luajit
/home/igalia/clopez/yocto-test/build_wpe/build_atlash/tmp/sysroots/x86_64-linux/usr/bin/luajit

Reply:
Notice also that ```$(shell pkg-config luajit --variable=prefix)``` will on the 99% of the cases just evaluate to /usr, because what we are getting here is the install prefix of the luajit for the **target** system.

But we don't want that at build time (because we are cross-building). We want the path to the luajit-native program on the build host. And the info of where that is located is not available at all on the pkg-config file.

So I believe the only sane thing we can do here is to not try to guess the path of the luajit program and just expect it to be on ```PATH```

Reply:
TL;DR: While I understand what you are trying to solve, I disagree on the method and I argue that the *saner* option (if not *the sanest*) is that each recipe explicitly uses full paths to the host tools that it uses.


---

(I'll use Buildroot terminology in this comment, so “host” is “machine that cross-builds” and “target” is “machine which executes the built product.)

Each build recipe that needs a host-version of LuaJIT should be setting the `LUAJIT` variable and making it point to the full path of the built host binary. For example in a `.mk` Buildroot file:

```makefile
# Needed.
MYPACKAGE_DEPENDENCIES += host-luajit

# Depending on the build system, MYPACKAGE_MAKE_ENV may be a better option
MYPACKAGE_MAKE_OPTS += LUAJIT='$(HOST_DIR)/bin/luajit'
MYPACKAGE_CONF_ENV += LUAJIT='$(HOST_DIR)/bin/luajit'
```

This is the preferred way of using tools build by `host-*` packages. For example you can see this pattern in the [qt5websockets.mk recipe](https://git.busybox.net/buildroot/tree/package/qt5/qt5websockets/qt5websockets.mk#n28), where the path to `qmake` is provided explicitly:

```makefile
define QT5WEBSOCKETS_CONFIGURE_CMDS
	(cd $(@D); $(TARGET_MAKE_ENV) $(HOST_DIR)/bin/qmake)
endef
```

Being explicit and making recipes avoids hard to debug build issues which usually go unnoticed, like:

- Build systems/scripts which modify `$PATH`.
- Using a system-wide binary instead of the one provided by a `host-*` dependency:
   - Missing dependencies on `host-*` packages may go unnoticed if not using the full path.
   -  System-wide binaries may be missing build options and/or patches applied by cross-compilation frameworks.

Overall, I think Buildroot does the right thing here, and those recipes that use programs provided by `host-*` packages *must* explicitly use full paths to the tools they use. The author of a build recipe *knows* when/if a tool provided by a `host-*` package is used, therefore they can perfectly add the full paths when writing it. Isn't it?

Last but not least, what about cross-compilation frameworks which do prepend the path of its host-tools to `$PATH`? In this case you would need anyway to set `LUAJIT=` to the full path of the tool anyway. So I don't see why not doing it always in a cross-compilation scenario, regardless of the cross-compilation framework being used, and avoid potential issues by being on the safest scenario possible (that is: the tool path points to the exact location where we know that the precise version we want to use must be).

Reply:
Sorry for the late reply ... 

I think this conversation is getting more complex than it should, and I'm not really interested in having here a discussion about build frameworks. I just want to fix a build issue that happens currently with dyz and OpenEmbedded. So, I will try to stick to the point.

My 2 points:

1. Specifying a specific path to the luajit binary with the environment variable ```LUAJIT``` is still possible after this patch. This patch doesn't change that. So if your buildroot recipe relies on this, that's fine. It will continue to work.
2. The only thing this patch changes is what the build system executes in case the ```LUAJIT``` environment variable is **not** defined:
    * Currently it is picking a default of ```LUAJIT=/usr/bin/luajit```
    * After this path it will default to just ```LUAJIT=luajit```
        * So it will allow to execute the ```luajit``` binary if its in any path of ```PATH``` other than just ```/usr/bin/```

So that's it.

Do you have any concern with this change of behaviour?

Reply:
Also I think this patch will allow building ```dyz``` on distributions that don't have a luajit binary on /usr/bin (like NixOS) or where the user installed luajit from sources (doing that will get with luajit installed in /usr/local/bin by default)

Reply:
In the case of NixOS, I would expect it to have the correct path in the `.pc` file, otherwise they probably need to update their build recipe for LuaJIT. Anyway, back into topic: I do not have a very strong opinion against merging this PR, I just feel like it's solving the issue in the wrong place. As long as an option to pass a `LUAJIT=` environment variable with the correct path, I'll be happy enough :wink: